include/caffe/util/external_lib_data_source.hpp

```diff
@@ -0,0 +1,51 @@
+/**
+* @brief    This file contains interface declarations that external data source
+*           needs to implement to be able to plug into caffe pipeline. External
+*           library needs to export the factory method with the signature
+*           defined below.
+*           To instruct caffe to use external lib as data source one needs to
+*           use external lib data layer.
+*/
+#ifndef EXTARNAL_LIB_DATA_SOURCE_H_
+#define EXTARNAL_LIB_DATA_SOURCE_H_
+
+class IExternalLibDataSource;
+
+// External library data source factory method signature.
+typedef IExternalLibDataSource* (*ExternalLibDataSourceFactoryMethod)(
+  const char* external_lib_params);
+
+enum BlobType { BlobTypeFLOAT, BlobTypeDOUBLE };
+
+/**
+* @brief    Defines interface for one data point.
+*/
+class IDatum {
+ public:
+  // Returns datum shape for the blob with the given name.
+  virtual void GetBlobShape(const char* blob_name, const int** shape,
+    int* shape_count) = 0;
+
+  // Returns data of the given type for the blob with the given name.
+  virtual void GetBlobData(const char* blob_name, void* blob_data,
+    BlobType type) = 0;
+};
+
+/**
+* @brief    Defines interface for data source.
+*/
+class IExternalLibDataSource {
```

Comment:
Why did you create additional interfaces? Caffe already has an interface for data layers: `BaseDataLayer<Dtype>`. Why not let the loaded module create instances of that class instead of the new interfaces? I'd expect that would save a lot of code overhead. On top of that I think it would also simplify the process for supporting compute layers.